### ISSUE TITLE
Broken BaseHtml::getAttributeValue()

### ISSUE THREAD

**Issue by desher** [yiisoft/yii2#12407]
### What steps will reproduce the problem?

This code based on http://www.yiiframework.com/doc-2.0/guide-input-file-upload.html and previously worked correctly.
Broken in https://github.com/yiisoft/yii2/commit/3fcc26996b9395d2892839b5bc8071c99a6cab07
Method getAttributeValue()

**Model**
```php
class Blog extends BlogFeedPostAbstract
{
    public static function tableName()
    {
        return 'blog';
    }

    public function rules()
    {
        return [
            [['files'], 'image', 'extensions' => 'png, jpg', 'minFiles' => 2, 'maxFiles' => 10],
        ];
    }

   ...
```

**Controller**
```php
...

public function actionIndex()
{
    $model = new Blog();

    if ($model->load(Yii::$app->request->post())) {

        // getInstances return array of files
        $model->files = UploadedFile::getInstances($model, 'files');

        if ($model->save()) {
            if ($model->files) {
                // save files..
            }
            $this->redirect(['index']);
        }
    }

    return $this->render('upload_form', [
        'model' => $model,
    ]);
}
```

**View**
```php

<?php $form = ActiveForm::begin([
    'options' => [
        'enctype' => 'multipart/form-data',
    ],
]); ?>

    <?= Html::activeFileInput($model, 'files[]', [
        'id' => 'upload-file',
        'multiple' => true,
        'accept' => 'image/*',
    ]); ?>
    
<?php ActiveForm::end(); ?>
```

We a submitting form with empty file input.

### What is the expected result?

Model has errors ('minFiles' => 2) and form rendered again.

### What do you get instead?

PHP Notice – yii\base\ErrorException
Array to string conversion

in /var/www/child/vendor/yiisoft/yii2/helpers/BaseHtml.php
line 537
```php
$options['value'] = $value === null ? null : (string) $value;
```

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.10-dev
| PHP version      | 7.0.8-0ubuntu0.16.04.2
| Operating system | Ubuntu 16.04

**Comment by dmdb** [comment]
Also getAttributeValue() method is hard to understand with such a code style. Why 
`$matches[3] !== '[]'` ?
```
if (!preg_match('/(^|.*\])([\w\.]+)(\[.*|$)/', $attribute, $matches)) {
    throw new InvalidParamException('Attribute name must contain word characters only.');
}
```
Why exception requires us to use only word characters while accepting []?
https://github.com/yiisoft/yii2/commit/3fcc26996b9395d2892839b5bc8071c99a6cab07#diff-95842e557ddaa8b8fa5d26cf31fefae7R2034

**Comment by SilverFire** [comment]
Failed to reproduce

> `'minFiles' => 2`

This property does not belong neither to the `ImageValidator` nor to the `FIleValidator`. Is this property your custom?

**Comment by SilverFire** [comment]
> Also getAttributeValue() method is hard to understand with such a code style. Why 
> $matches[3] !== '[]' ?

Regular expressions are often hard to understand :)

> Why exception requires us to use only word characters while accepting []?

`[]` means that attribute will be represented as array.

**Comment by desher** [comment]
> This property does not belong neither to the ImageValidator nor to the FIleValidator. Is this your custom property?

Yes, this is own custom property. But it is not important for this issue:)
Exception *Array to string conversion* throws by BaseHtml for attribute files[] after $model->files = UploadedFile::getInstances($model, 'files');

**Comment by desher** [comment]
Previously BaseHtml::getAttributeValue() returns null for `files[]` attribute. Now this returned empty array.

**Comment by SilverFire** [comment]
Verified as a BC breaking, thank you for a report.

However, the previous behavior was not consistent:

```php
$model->foo = [];

var_dump(Html::getAttributeValue($model, 'foo')); // array[]; seems to be expected
var_dump(Html::getAttributeValue($model, 'foo[]')); // null; same here, element with key [] (empty key) is not present in array `foo`

$model->foo = ['type1', 'type2', ['sub-array', 'sub-array2']];

var_dump(Html::getAttributeValue($model, 'foo[1]')); // string(5): "type2"; expected, `$foo[1]` exists and equals string "type2"
var_dump(Html::getAttributeValue($model, 'foo[1][]')); // string(5): "type2"; wtf, why not `null`?
```

The problem seems to be not so obvious. @yiisoft/core-developers what do you think

**Comment by samdark** [comment]
Hmm... logically it should be `array` in this case. I'm for fixing all the usage of `Html::getAttributeValue()` and adding notes to UPGRADE.

**Comment by cebe** [comment]
There are a lot of use cases for this method and it will break a lot of things changing the behavior of it. Imo we should cover all use cases in test cases and then make the behavior logical and consistent in 2.1.

**Comment by SilverFire** [comment]
Original changes were reverted https://github.com/yiisoft/yii2/commit/c66b845f4809125d3e320ce138eed3b1ecb7b39d because of BC breaking and no change to normalize behavior without BC.
